More RISC-V instructions in `core::arch`

luojia65 opened the pull request:
This pull request includes:

- Hypervisor extension, including memory management insturctions and virtual machine memory load and store instructions
- Supervisor memory management, including address translation fence insturctions (important for Rust kernel development) and cache invalidation insturctions
- Non privileged instructions FENCE.I, NOP and WFI

Module `riscv64` is added to include RV64 only virtual machine memory instructions. RV64 supports all RV32 virtual machine instructions, it's implemented like:

```rust
    pub mod riscv64 {
        pub use crate::core_arch::riscv::*;
        pub use crate::core_arch::riscv64::*;
    }
```

(Edit: implemented like `pub use crate::core_arch::riscv32::*;` etc)

Most of implemented functions are marked as unsafe. That's because these functions either load or store effectively into raw pointer (hlv, hlvx and hsv instructions), or would effect how the RISC-V core handles page memory translation (memory management fences and cache invalidations).

Specifically, here's a list of implemented instruction functions from updated RISC-V core arch module:

<details>

![图片](https://user-images.githubusercontent.com/40385009/147846592-c5576dab-e6d7-405d-a6b0-cd2b3630c752.png)
</details>

~~Marked as draft: more polishing on documents~~(done)~~, and decide whether #[must_use] is necessary on virtual machine memory loads.~~(ref: ptr::read & ptr::write, does not put #[must_use] by now)

rust-highfive commented:
r? @Amanieu

(rust-highfive has picked a reviewer for you, use r? to override)

Amanieu commented on crates/core_arch/src/mod.rs:
```diff
@@ -66,6 +66,17 @@ pub mod arch {
         pub use crate::core_arch::riscv::*;
```
If we're going to have separate intrinsics for riscv32 and riscv64 then we should probably rename the existing `riscv` module to `riscv32`. This makes it consistent with the other architectures.

luojia65 replied:
Yes, I'd accept this suggestion. RISC-V RV64 only contains more instructions in each set than RISC-V RV32, no instructions are removed from RISC-V RV32 yet. RV32 is not a legacy extension, it's better to make it explict. This way provides more readable rust docs as well.
After this change, I'd modify the implementation of `core::hint::spin_loop` in standard library.

Amanieu reviewed (commented):
LGTM other than the module name issue.

luojia65 commented:
I modified `riscv` to `riscv32`.

r? @Amanieu

Amanieu commented on crates/core_arch/src/riscv32/mod.rs:
```diff
@@ -0,0 +1,481 @@
+/// advancing the `pc` and incrementing any applicable performance counters.
+#[inline]
+pub fn nop() {
+    unsafe { asm!("nop") }
```
All of the `asm!` blocks here should use `nostack` since they don't push anything to the stack.

Additionally:
- Most should use `nomem`, except for loads which should use `readonly` and stores which should use neither.

luojia65 replied:
Thanks, I'll check it out.
There are many privileged memory fences as well, I'd also figure it out in these intrinsic functions.

Amanieu replied:
Good point about the fences, those should not use `nomem`/`readonly` because they need to act as a compiler fence.

Amanieu commented on crates/core_arch/src/mod.rs:
```diff
@@ -264,7 +276,11 @@ mod arm;
 #[cfg(any(target_arch = "riscv32", target_arch = "riscv64", doc))]
 #[doc(cfg(any(target_arch = "riscv32", target_arch = "riscv64")))]
-mod riscv;
+mod riscv32;
```
Perhaps calling this `riscv_shared` (like `arm_shared`) makes more sense?

luojia65 replied:
That makes sense, it would be a more common module system design.

luojia65 commented:
Hello! I renamed internal shared riscv module to `riscv_shared`, and modified asm! options under following rules:

- Trivial no-op instruction (NOP), wait-for-interrupt (WFI): `options(nomem, nostack)`
- Unprivileged fences (FENCE.I), privileged fences (SFENCE.*, SINVAL.*, HFENCE.*, HINVAL.*): `options(nostack)`
- Hypervisor loads (HLV.*, HLVX.*): `options(readonly, nostack)`
- Hypervisor stores (HSV.*): `options(nostack)`

r? @Amanieu

luojia65 commented on crates/core_arch/src/riscv64/mod.rs:
```diff
@@ -0,0 +1,49 @@
+/// This operation is not available under RV32 base instruction set.
+///
+/// This function is unsafe for it accesses the virtual supervisor or user via a `HLV.WU`
+/// instruction which is effectively an unreference to any memory address.
```
note: in following pull requests the word 'unreference' should be replaced by 'dereference' for consistency

luojia65 replied:
fixed in https://github.com/rust-lang/stdarch/pull/1278
